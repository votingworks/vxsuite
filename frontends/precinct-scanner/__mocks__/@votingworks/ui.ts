// eslint-disable-next-line import/no-extraneous-dependencies, import/no-import-module-exports
import {
  fakePrintElementWhenReady,
  fakePrintElement,
} from '@votingworks/test-utils';

const ui = jest.requireActual('@votingworks/ui');

module.exports = {
  ...ui,
  printElement: fakePrintElement,
  printElementWhenReady: fakePrintElementWhenReady,
};
