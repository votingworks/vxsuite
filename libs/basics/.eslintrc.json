{
  "extends": ["plugin:vx/recommended"],
  "rules": {
    // disable this rule since this package tests `Result` and needs
    // to have tests that assert on the result's predicates
    "vx/no-assert-result-predicates": "off"
  }
}
