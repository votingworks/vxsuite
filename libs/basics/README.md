# basics

An extended TS standard library of generic utility functions and types.

This library should only include utilities that are not specific to VotingWorks
products, but would be generally applicable to any TypeScript project (similar
to a library like lodash). In other words, it should be considered as an
extension of the standard TS libraries. This means it shouldn't depend on any
other monorepo packages (which will help avoid dependency cycles). Anything
complex or specific to a particular domain should live in its own library.
