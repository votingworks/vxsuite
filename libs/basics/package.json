{
  "name": "@votingworks/basics",
  "version": "1.0.0",
  "private": true,
  "description": "An extended TS standard library of generic utility functions and types.",
  "license": "GPL-3.0-only",
  "main": "build/index.js",
  "types": "build/index.d.js",
  "scripts": {
    "build": "pnpm --filter $npm_package_name... build:self",
    "build:self": "tsc --build tsconfig.build.json",
    "clean": "pnpm --filter $npm_package_name... clean:self",
    "clean:self": "rm -rf build && tsc --build --clean tsconfig.build.json",
    "lint": "pnpm type-check && eslint .",
    "lint:fix": "pnpm type-check && eslint . --fix",
    "pre-commit": "lint-staged",
    "test": "is-ci test:ci test:watch",
    "test:ci": "pnpm build && pnpm test:coverage --run",
    "test:coverage": "vitest --coverage",
    "test:watch": "vitest",
    "type-check": "tsc --build"
  },
  "dependencies": {
    "deep-eql": "4.1.3",
    "util": "^0.12.4"
  },
  "devDependencies": {
    "@types/deep-eql": "^4.0.2",
    "@types/node": "20.16.0",
    "@vitest/coverage-istanbul": "^2.1.8",
    "eslint-plugin-vx": "workspace:*",
    "fast-check": "2.23.2",
    "is-ci-cli": "2.2.0",
    "lint-staged": "11.0.0",
    "sort-package-json": "^1.50.0",
    "vitest": "^2.1.8"
  },
  "packageManager": "pnpm@8.15.5"
}
