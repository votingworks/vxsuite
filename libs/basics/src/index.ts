/* istanbul ignore file */
export * from './assert';
export * as collections from './collections';
export * from './date_without_time';
export * from './deferred';
export * from './duplicates';
export * from './group_by';
export * from './equality';
export * from './errors';
export * from './find';
export * from './iterators';
export * from './map_object';
export * from './merge_objects';
export * from './result';
export * from './range';
export * from './sleep';
export * from './typed_as';
export * from './types';
export * from './type_guards';
export * from './unique';
