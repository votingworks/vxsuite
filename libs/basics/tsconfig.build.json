{
  "extends": "./tsconfig.json",
  "include": ["src"],
  "exclude": ["**/*.test.ts"],
  "compilerOptions": {
    "noEmit": false,
    "rootDir": "src",
    "outDir": "build",
    "declaration": true,
    "declarationMap": true
  },
  "references": [{ "path": "../../libs/eslint-plugin-vx/tsconfig.build.json" }]
}
