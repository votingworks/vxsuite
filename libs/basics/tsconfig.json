{
  "extends": "../../tsconfig.shared.json",
  "include": ["src"],
  "exclude": [],
  "compilerOptions": {
    "module": "node16",
    "lib": ["esnext"],
    "composite": true,
    "noEmit": true,
    "esModuleInterop": true,
    "skipLibCheck": true
  },
  "references": [{ "path": "../../libs/eslint-plugin-vx/tsconfig.build.json" }]
}
