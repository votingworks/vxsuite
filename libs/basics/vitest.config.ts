import { defineConfig } from 'vitest/config';

export default defineConfig({
  test: {
    coverage: {
      thresholds: { 100: true },
      provider: 'istanbul',
      include: ['src/**/*.ts'],
    },
  },
});
