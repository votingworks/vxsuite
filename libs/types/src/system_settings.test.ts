import { assert } from '@votingworks/basics';

import {
  DEFAULT_SYSTEM_SETTINGS,
  safeParseSystemSettings,
} from './system_settings';

const systemSettingsString = JSON.stringify(DEFAULT_SYSTEM_SETTINGS);

test('safeParseSystemSettings safely parses valid system settings', () => {
  const parsed = safeParseSystemSettings(systemSettingsString);
  assert(parsed.isOk());
  expect(parsed.unsafeUnwrap()).toEqual(DEFAULT_SYSTEM_SETTINGS);
});

test('safeParseSystemSettings returns an error for malformed system settings', () => {
  const parsed = safeParseSystemSettings(JSON.stringify({ invalid: 'field' }));
  assert(parsed.isErr());
});

test('disallows invalid mark thresholds', () => {
  expect(
    safeParseSystemSettings(
      JSON.stringify({
        ...DEFAULT_SYSTEM_SETTINGS,
        markThresholds: { definite: 0.2, marginal: 0.3 },
      })
    ).unsafeUnwrapErr()
  ).toMatchSnapshot();

  expect(
    safeParseSystemSettings(
      JSON.stringify({
        ...DEFAULT_SYSTEM_SETTINGS,
        markThresholds: { marginal: 0.3 },
      })
    ).unsafeUnwrapErr()
  ).toMatchSnapshot();

  expect(
    safeParseSystemSettings(
      JSON.stringify({
        ...DEFAULT_SYSTEM_SETTINGS,
        markThresholds: { definite: 1.2, marginal: 0.3 },
      })
    ).unsafeUnwrapErr()
  ).toMatchSnapshot();
});

test('disallows invalid adjudication reasons', () => {
  expect(
    safeParseSystemSettings(
      JSON.stringify({
        ...DEFAULT_SYSTEM_SETTINGS,
        precinctScanAdjudicationReasons: ['abcdefg'],
        centralScanAdjudicationReasons: ['abcdefg'],
      })
    ).unsafeUnwrapErr()
  ).toMatchSnapshot();

  expect(
    safeParseSystemSettings(
      JSON.stringify({
        ...DEFAULT_SYSTEM_SETTINGS,
        centralScanAdjudicationReasons: 'foooo',
      })
    ).unsafeUnwrapErr()
  ).toMatchSnapshot();
});
