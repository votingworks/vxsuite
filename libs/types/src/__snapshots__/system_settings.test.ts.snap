// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`disallows invalid adjudication reasons 1`] = `
[ZodError: [
  {
    "received": "abcdefg",
    "code": "invalid_enum_value",
    "options": [
      "UninterpretableBallot",
      "MarginalMark",
      "Overvote",
      "Undervote",
      "BlankBallot",
      "UnmarkedWriteIn"
    ],
    "path": [
      "centralScanAdjudicationReasons",
      0
    ],
    "message": "Invalid enum value. Expected 'UninterpretableBallot' | 'MarginalMark' | 'Overvote' | 'Undervote' | 'BlankBallot' | 'UnmarkedWriteIn', received 'abcdefg'"
  },
  {
    "received": "abcdefg",
    "code": "invalid_enum_value",
    "options": [
      "UninterpretableBallot",
      "MarginalMark",
      "Overvote",
      "Undervote",
      "BlankBallot",
      "UnmarkedWriteIn"
    ],
    "path": [
      "precinctScanAdjudicationReasons",
      0
    ],
    "message": "Invalid enum value. Expected 'UninterpretableBallot' | 'MarginalMark' | 'Overvote' | 'Undervote' | 'BlankBallot' | 'UnmarkedWriteIn', received 'abcdefg'"
  }
]]
`;

exports[`disallows invalid adjudication reasons 2`] = `
[ZodError: [
  {
    "code": "invalid_type",
    "expected": "array",
    "received": "string",
    "path": [
      "centralScanAdjudicationReasons"
    ],
    "message": "Expected array, received string"
  }
]]
`;

exports[`disallows invalid mark thresholds 1`] = `
[ZodError: [
  {
    "code": "custom",
    "message": "marginal mark threshold must be less than or equal to definite mark threshold",
    "path": [
      "markThresholds"
    ]
  }
]]
`;

exports[`disallows invalid mark thresholds 2`] = `
[ZodError: [
  {
    "code": "invalid_type",
    "expected": "number",
    "received": "undefined",
    "path": [
      "markThresholds",
      "definite"
    ],
    "message": "Required"
  }
]]
`;

exports[`disallows invalid mark thresholds 3`] = `
[ZodError: [
  {
    "code": "too_big",
    "maximum": 1,
    "type": "number",
    "inclusive": true,
    "exact": false,
    "message": "Number must be less than or equal to 1",
    "path": [
      "markThresholds",
      "definite"
    ]
  }
]]
`;
